docs/data/date-pickers/custom-field/BrowserV7SingleInputRangeField.js

```diff
@@ -1,7 +1,7 @@
 import * as React from 'react';
 
 import useForkRef from '@mui/utils/useForkRef';
-import { useSlotProps } from '@mui/base/utils';
+import useSlotProps from '@mui/utils/useSlotProps';
```

Comment:
This feels like it goes against https://www.notion.so/mui-org/mui-utils-purpose-9a9fc9da3a004864b6c4e1f4d1f24f95. Why not do:

- We make `@mui/utils` private. It's owned by Base UI. Only Pigment CSS is allowed to depend on it
- We expose most legacy APIs of `@mui/base` in `@base_ui/react`
- We deprecate `@mui/base`

Reply:
We want to remove the dependency on @mui/base ASAP, it's hard to maintain it, especially because Material UI depends on it (both on master and next). This is an intermediate step towards that. More over there are things used in Mateiral UI and MUI X that will never live in the new Base UI. This util is a great example (`useSlotProps`), as Base UI will not have the notion of slots, it's wrong to add it as a logic there, just because Material UI and MUI X depend on it. I see value in having @mui/utils that serves the purpose as a common utils package. 

Considering the above, this statement does not make sense to me:
> Should we re-export all of it from @mui/base? If yes, does it even make sense to keep the utils package?

There are MUI X packages that do not depend on Material UI, so these common utils have to live somewhere to avoid duplication.

Reply:
I think we have a discussion to have about where those utility methods should live when X needs them.
Because so far I read that:

- maybe X should not use `@mui/utils`
- `@mui/base` is going away at some point
- X should not depend on `@mui/material` in the future so that we truly support several design systems

This leaves us with `@base_ui` (maybe `@base_ui/utils`), but as @mnajdova I find it super weird to put stuff in it that talks about concepts `@base_ui` doesn't even have.

I don't like `@mui/utils` as it is handled right now, because we sometimes use it in the docs, the name doesn't say that people should not use it, all the imports are unstable **EXCEPT** if you import from the nested endpoints.
But I do think we need an internal package (with a name that clearly states it's internal, with linting rules that block us from using it in the doc) to share code at a company level. And if we decide that we don't need it, then we acknowledge that X will just duplicate those utils in `@mui/x-internals`.

Reply:
Which @mui/utils functions are used in the demos? On Core, we have: 
- visuallyHidden - it's stable
- deepmerge - it's stable
- capitalize - not stable
- useForkRef - not stable

Maybe it's worth exporting these from @mui/material and @mui/x* respectively, and we can keep @mui/utils as internal package.

Reply:
> Which @mui/utils functions are used in the demos? On Core, we have:

- `useForkRef` a lot
- `useEnhancedEffect` on some occasions
- `useId` on some occasions
- `debounce` on some occasions
- `useControlled` on one occasion 

And with this PR we will start using:

- `useSlotProps` a lot on custom field demos (because if you override the `field` slot, you still need to resolve the `textField` slot inside it. It's not great but I don't have a clear vision on how to avoid that.

For stuff like `useId`, `useForkRef`, `capitalize`, `debounce`, I think we can either:

1. Export them from `@mui/x*` packages if we need them in the X doc (would have to export them from each X package, it would probably look like `@mui/x-date-grid/utils`, `@mui/x-tree-view/utils`, etc...)
2. Split the public and private utils (private utils in `@mui/private`, public utils in `???`) and keep importing from a share util package
3. Inline in the doc (for `useEnhancedEffect` it can make sense, for stuff like `visuallyHidden` or `deepmerge` I don't think it's a good idea)

Reply:
Given the the fact that we haven't landed on a decision in https://www.notion.so/mui-org/mui-utils-purpose-9a9fc9da3a004864b6c4e1f4d1f24f95 and this comment:
> It is not required to merge this PR, everything that you had before will still work, however we don't plan to continue to do releases of Base UI, all fixes are going to be done directly in @mui/utils and @mui/material. I will leave it up to the MUI X team to decide on what's best as a next step.

I would vote to firstly finalize the decision of the `@mui/utils` package.
Given the decision—proceed with making necessary changes (if any) and then adjust this PR to drop the `@mui/base` dependency.

`@mui/base` has been there for quite some time and seems stable, I don't see a big rush in removing it before a new major if we do not see any issues that MUI Core can't (don't want to) fix. 🤔 

WDYT @flaviendelangle @mnajdova @oliviertassinari @mui/x?

Reply:
I gave it more thought and am proposing to proceed with this change.
In a follow we can re-export anything that is used in docs from `@mui/utils` via the `<package>/utils` folder to avoid the need for end users to install `@mui/utils` in turn hiding its usage from end users. 🤔 
Other than the open questions in regards to dep versions this looks good to me. 👌

Reply:
A PR with the above ☝️ idea implementation: https://github.com/mui/mui-x/pull/13901

Reply:
> We want to remove the dependency on @mui/base ASAP, it's hard to maintain it, especially because Material UI depends on it (both on master and next)

@mnajdova I don't follow the reasoning. I assume that Material UI depending on Base UI having the source in a separate repository is the source of what makes it "hard". But isn't it something we can't get away from? So shouldn't we as soon as possible embrace this challenge, and start to build around this new constraint? It feels like we are dodging it in the short term to eventually have a hard time in the future. Eventual integration doesn't work, so it feels like a step backward.

Reply:
> https://www.notion.so/mui-org/mui-utils-purpose-9a9fc9da3a004864b6c4e1f4d1f24f95

I have updated the proposed Option A to account for the new repository split and Base UI npm package name.

> I honestly don't understand why we would go through all this trouble of adding stuff in @base_ui/react that we don't even know if @base_ui/react will use it.

For shared utils only used in MUI X, having shared utils makes sense. I imagine it's why we created @mui/x-internals`.

What confuses me is that at its root, Base UI is about providing the primitives to create a whole set of components. It feels like `useSlotProps` is in this product scope. Something that MUI X **and**  other community package extensions can then use.

We could have it in MUI X, but I don't really see https://github.com/KevinVandy/material-react-table importing it from MUI X.

Reply:
I'm not proposing the put `useSlotProps` in MUI X
For me `useSlotProps` should remain in `@mui/utils` until we know if Base UI will use it in its exact current shape.
Once MUI X uses the same APIs as Base UI, we will be able to get rid of `@mui/utils`, but moving stuff to `@base-ui/react` before we know what the API of the Base UI slots will look like is IMHO a mistake.

Reply:
@flaviendelangle Yeah, OK. I have updated the proposal in https://www.notion.so/mui-org/engineering-mui-utils-purpose-9a9fc9da3a004864b6c4e1f4d1f24f95?pvs=4#d1699bb48f0f4ef58ca831d1ee9772ec for the helper functions. In that mode, we would grow to have a `@mui/utils` package as a public API that we document. 

`useSlotProps` still feels more Base UI product scope than Material UI product scope to me, but no strong preferences.

Reply:
> `useSlotProps` still feels more Base UI product scope than Material UI product scope to me, but no strong preferences.

But that's still not my point :thinking: 
I fully agree that at some point MUI X should consume this utility function from `@base-ui/react` (if Base UI have a notion of slots) and not depend on Material UI packages at all. 

My point is that if `useSlotProps` is migrated to `@base-ui/react`, then we must be almost certain that when Base UI starts having slots (for their single component approach), they will use `useSlotProps` in its current shape (same params, same generics, etc...).

Otherwise, we would create our self a painful migration that we could easily avoid by just not putting `useSlotProps` in `@base-ui/react` for now. If MUI X uses some Base UI components by the time Base UI changes `useSlotProps`, then we could not update `@base-ui/react` without first migrating all of the `useSlotProps` to the new DX (which could potentially involve breaking changes if things like the `ownerState` are not handled the same way).

I like the idea of having Base UI building its new tools and the other projects migrate to it bits by bits, rather than copy paste some of the existing tools in Base UI, hoping that they are how Base UI will need them to be.
And for me it's far from obvious, we already have a variant of `useSlotProps` in Joy UI (`useSlot`) that don't have the same signature at all (I did not dive deeper into why tbh).

Reply:
> then we could not update @base-ui/react without first migrating all of the useSlotProps to the new DX (which could potentially involve breaking changes if things like the ownerState are not handled the same way).

@flaviendelangle I would expect a transition to be incremental, that we would add a new API, first, and only after deprecating the old one.

> I like the idea of having Base UI building its new tools and the other projects migrate to it bits by bits, rather than copy paste some of the existing tools in Base UI

I see this path with some weaknesses though:

- How do we create a smooth migration experience? How can a product owner of Base UI broadcast effectively to everyone that A is the new way to do B? How can people doing B can review A before it's merged before it moves to be stable and find holes? On the other hand, we could already slice the product scope, and set ownership areas. 
- How do we ensure coherence in the product vision? How do we flag that B is meant to be replaced by A in Base UI.

Now, for public APIs, e.g. Portal, to me, it feels clear, Base UI. Then, there needs to be a discussion on the value of the abstraction vs. using React.createPortal directly, e.g.

<img width="864" alt="SCR-20240721-qjmp" src="https://github.com/user-attachments/assets/297eee87-992e-4f0f-8201-3a2ed4871662">

https://2023.stateofreact.com/en-US/features/ could point toward; it should be public 

For `useSlotProps()`, It looks like it needs to be a public API because you can't truly extend a Material UI component without, MUI X being the proof, but is not very public, so ok, maybe we can rush this more.

Reply:
> @mnajdova I don't follow the reasoning. I assume that Material UI depending on Base UI having the source in a separate repository is the source of what makes it "hard". But isn't it something we can't get away from? So shouldn't we as soon as possible embrace this challenge, and start to build around this new constraint? It feels like we are dodging it in the short term to eventually have a hard time in the future. Eventual integration fail almost all the time, so it feels like a step backward.

No, I was referring to having two versions of @mui/base while it is unstable and having to release it from both master & next if changes are required. It doesn't have anything with it being in a separate repository. We didn't want to bump the version of @mui/base to v6 on the next branch, so ended in a situation where it is v5 in both master and next, and the changes released got out of sync. We decided in the end to release only from next, but things got us blocked into releasing some fixes on master (v5), like some Textarea changes.


> I would expect that we spend the same energy into depending on @base_ui/react. Like the last release was two months ago, start to make more frequent releases to be able to start to dogfood it quietly. @mui/base has a good number of downloads, this user base to need to have a migration flow possible to @base_ui/react, and not to @mui/material.

I don't understand how is this related. We are not going to remove the package or do anything similar. It's just about removing the internal dependency on this package. Again, I am not saying this is critical and urgent, I am just saying that we don't plan to do future releases of the package, as the changes were already moved to @mui/material, so we are going to do fixes directly in Material UI. Again, the changes are not urgent, I just thought I would do them here while I was doing them in Material UI, to see if there is anything missing that we need to re-export.

----

On the utils being imported from @base_ui/react, once we have the utils and the Base UI teams decide which ones makes sense as a scope of Base UI, we can update the imports. To me, just felt that some of them, `useSlotProps` as an example, will likely not live there. I didn't want to block this progress, by waiting on this decision, we already have a shared utils package for this (@mui/utils).

Reply:
@oliviertassinari @mnajdova @michaldudak @flaviendelangle 

> Base UI will not have the notion of slots, it's wrong to add it as a logic there
> I find it super weird to put stuff in it that talks about concepts @base_ui doesn't even have.
> I have heard talks about having single component exported from Base UI themes too.
> useSlotProps still feels more Base UI product scope than Material UI product scope to me, but no strong preferences.

Base UI will never have the notion of slotProps, it is not compatible with the new API design, and not something we will ever support or recommend. Even if we hypothetically shipped Base UI Themes (presumably a styled lib that Base UI team owns?), that would not use slotProps either.

> unstable does not mean internal, so adding it there and say that maybe we will remove it in the future is not a good approach DX-wise IMHO.

I agree.

> I think @mui/material should trend toward only, yes only, have CSS logic exported.

Styled libs should provide more than just CSS on top of unstyled libs. They should provide things like higher-level APIs (wrapping Base UI), setting prop defaults, additional components that Base UI doesn't export, JS helpers for handling things like generating automated breakpoints etc. slotProps are a good example of that. If a styled lib wants to provide an API like slotProps, that is their responsibility. It has nothing to do with Base UI. If multiple styled libs that we own want to provide a slotProps API, then maybe it should be abstracted to a shared repo, but again, it has nothing to do with Base UI.

> What confuses me is that at its root, Base UI is about providing the primitives to create a whole set of components. It feels like useSlotProps is in this product scope. Something that MUI X and other community package extensions can then use.

This seems like a fundamental misalignment on Base UI’s purpose. Base UI's purpose is not to provide Material UI, Joy UI, and X with everything they need, so that those libs can have the smallest surface area possible. Base UI's purpose is to provide every development team on Earth with a base layer of accessible components that they can _build on top of_ in order to create their product or design system.

Yes, some utils (like visuallyHidden, Portal etc.) can make sense in Base UI. But Base UI should not concern itself with high-level consumer APIs like slotProps.

> @mui/base has a good number of downloads, this user base to need to have a migration flow possible to @base_ui/react

Yes we need to safely migrate @mui/base users to @base_ui/react. We can manage this process via a notice in @mui/base docs, a migration doc, Discord support, and calls with larger users.

---

So in summary, I think we should move @mui/base utils to the most appropriate repo, so the teams who depend on them own the code, and are responsible for it. In most cases, that will not be Base UI.

Another example would be the `TextareaAutosize` util in @mui/base. There is an [issue in Base UI's repo](https://github.com/mui/base-ui/issues/167) about it. But the new Base UI lib does not have a Textarea component, and it's not roadmapped. I want to add it add some point (and support char count etc.), but it is not a critical component for Base UI and we plan to launch alpha without it.

If Material UI needs this util/component, then it should be Material UI's responsibility.

Reply:
We have a dedicated notion page to discuss where each util would live. For now, all of the utils used in Mateiral UI/MUI X were moved to @mui/utils. It's wrong to keep advertising @mui/base, when we are doing a complete re-write, so I honestly don't see how this can be a step back. Eventually, I agree that maybe we can move some of these to @base_ui. Once we decide on it, we can update the usages.